### ISSUE TITLE
direct mode: pfn: no seed

### ISSUE THREAD

**Issue by pohly** [intel/pmem-csi#480]
Recently (https://cloudnative-k8sci.southcentralus.cloudapp.azure.com/job/pmem-csi/view/change-requests/job/PR-477/27/execution/node/104/log/?consoleFull for https://github.com/intel/pmem-csi/pull/477, https://cloudnative-k8sci.southcentralus.cloudapp.azure.com/blue/organizations/jenkins/pmem-csi/detail/PR-478/1/pipeline for https://cloudnative-k8sci.southcentralus.cloudapp.azure.com/blue/organizations/jenkins/pmem-csi/detail/PR-478/1/pipeline) we started to get intermittent failures:

```
14:16:38  pmem-csi-controller-0/pmem-driver: W1128 13:16:38.702163       1 controllerserver-master.go:237] failed to create volume name:pmem-csi-clear-worker2 id:sanity-controller-create-twice-5AB71DB6-10958FB2 on 89bcd0f0a3119dd538e3cf25e6111d3267bfb503: rpc error: code = Internal desc = Node CreateVolume: failed to create volume: pfn: no seed
...
[91m[1m[Fail] [0m[90mE2E [0m[0m[Driver: pmem-csi] [0m[90m[Testpattern: inline ephemeral CSI volume] ephemeral [0m[91m[1m[It] should support two pods which share the same volume [0m
15:22:16  [37m/mnt/workspace/pmem-csi_PR-477/vendor/k8s.io/kubernetes/test/e2e/storage/testsuites/ephemeral.go:259[0m
15:22:16  
15:22:16  [91m[1m[Fail] [0m[90mE2E [0m[0m[Driver: pmem-csi] [0m[90m[Testpattern: inline ephemeral CSI volume] ephemeral [0m[91m[1m[It] should support multiple inline ephemeral volumes [0m
15:22:16  [37m/mnt/workspace/pmem-csi_PR-477/vendor/k8s.io/kubernetes/test/e2e/storage/testsuites/ephemeral.go:259[0m
15:22:16  
15:22:16  [91m[1m[Fail] [0m[90mE2E [0m[0m[Driver: pmem-csi] [0m[90m[Testpattern: inline ephemeral CSI volume] ephemeral [0m[91m[1m[It] should create read-only inline ephemeral volume [0m
15:22:16  [37m/mnt/workspace/pmem-csi_PR-477/vendor/k8s.io/kubernetes/test/e2e/storage/testsuites/ephemeral.go:259[0m
```

Somehow this seems to happen only for ephemeral volumes? Could it be size related?

**Comment by pohly** [comment]
From https://github.com/intel/pmem-csi/pull/478#issuecomment-559723566
> We fail to get pfn seed device pointer to the given region, pfn holds some metadata of fsdax namespace. According to nvdimm documentation persistence of the PFN metadata is not important; this is just convenient because it scales with the persistent memory capacity. But I am not sure what are the side effects if we ignore these failures and use such namespace.

@avalluri: can you clarify with our colleagues?

**Comment by pohly** [comment]
What do we do with the PFN metadata?

**Comment by avalluri** [comment]
As `ndctl` code does, while creating a `fsdax` namespace, we get the `pnf` seed device of the region, and set mapping location and alignment of device and set the namespace of this pfn to this newly created namespace device.

Bun in this failure case, somehow we are not able to get the region's `pfn` device, which needs to be investigated.

**Comment by okartau** [comment]
I did some trial test runs in local govm-based cluster. The problem is not ephemeral-specific, and not Fedora-specific, as it also happens for regular volumes, and as well on govm-Ubuntu-based systems. The log files dont show root cause. After series of successful Volume Creations, another attempt fails with "pfn: no seed". There is no error-indicating kernel messages close to that point, and no resources shortage.
Next, I try to go back to 32G pmem media size and see does it make a difference.
Is it so that change 32G -> 64G was made to accommodate `-useforfsdax` percentage change to smaller?
As `-useforfsdax` size setting only affects LVM mode, direct mode is kind of side-victim of that change (if the size is part of problem here).

**Comment by pohly** [comment]
Olev Kartau <notifications@github.com> writes:
> I did some trial test runs in local govm-based cluster. The problem is
> not ephemeral-specific, and not Fedora-specific, as it also happens
> for regular volumes, and as well on govm-Ubuntu-based systems. The log
> files dont show root cause. After series of successful Volume
> Creations, another attempt fails with "pfn: no seed". There is no
> error-indicating kernel messages close to that point, and no resources
> shortage.

What's the underlying reason in the code for this error? Can it be
traced to the kernel failing in some way?

> Is it so that change 32G -> 64G was made to accommodate `-useforfsdax`
> percentage change to smaller?

Yes. By doubling the PMEM size and taking half of it for LVM, the size
of the LVM physical volume stayed the same.

**Comment by pohly** [comment]
Is this perhaps related?

https://patchwork.kernel.org/patch/11148999/

**Comment by okartau** [comment]
No idea about that one.
But with reverted media size to 32GB, I did run 10/10 e2e rounds in direct mode without seeing any "pfn: no seed" error.

**Comment by okartau** [comment]
About recent change of "use for" options: There is now combination in test/setup-deployment.sh:
```
value: "--useforfsdax=50"
value: "--useforsector=10"
```
and size was changed 32G->64G.
Is it needed to leave 40% unused, or can we leave smaller unused?
Is sector-mode space used in testing?
Assuming most of space is used for fsdax mode, can we arrange these fractions for example 90 and 5,  keeping the previous 32G size, and try to avoid "pfn: no seed" errors which seem te get triggered on 64G media.
True that this is workaround and not about finding real fix, but still, if we just leave space unused, then is it really needed to be 40%

**Comment by avalluri** [comment]
> Is it needed to leave 40% unused, or can we leave smaller unused?

This space is used for device manager direct mode unit tests. 

> Is sector-mode space used in testing?

No, as of i know.

> Assuming most of space is used for fsdax mode, can we arrange these fractions for example 90 and 5, keeping the previous 32G size, and try to avoid "pfn: no seed" errors which seem te get triggered on 64G media.

But is this(increased size) the root cause? For a quick test, I would suggest, make fsdax=50%, sector=0%, so that it is almost the same as the previous setup.
 
> f we just leave space unused, then is it really needed to be 40%

Not really, So far the tests try to create max four 16MB sized devices.

**Comment by pohly** [comment]
Olev Kartau <notifications@github.com> writes:
> But with reverted media size to 32GB, I did run 10/10 e2e rounds in
> direct mode without seeing any "pfn: no seed" error.

That can't be the solution. We need to find the root cause and address
it.

**Comment by okartau** [comment]
@avalluri expressed idea to also test without  ./pkg/pmem-device-manager in TEST_PKGS.
So I created a clean cluster with 64 G media on nodes, and ran 10x e2e, and result is 10x OK without any "pfn: no seed" errors.
Then I wanted to cross-check, so I added pmem-device-manager in back  in TEST_PKGS, and got pfn seed error right on next run in same cluster. Means, this also seems like a trigger. 
Overall it looks slightly confusing now, as it seems I can avoid pfn-seed errors in 2 ways: either by reducing media to 32 G or by excluding pmem-device-manager tests.
pmem-device-manager tests sequence somehow enables path to more error states? and these get even more exposed on bigger media? This is not clear at all.

**Comment by pohly** [comment]
Olev Kartau <notifications@github.com> writes:
> Overall it looks slightly confusing now, as it seems I can avoid pfn-seed errors in 2 ways: either by reducing media to 32 G or by excluding pmem-device-manager tests.
> pmem-device-manager tests sequence somehow enables path to more error
> states? and these get even more exposed on bigger media? This is not
> clear at all.

IMHO we need to dig deeper to find the root cause. Where does the error
come from? Which component is not behaving as expected?

Experimenting may be useful, but as you found out, the results do not
explain why it happens.

**Comment by okartau** [comment]
You are right, let's dig in. Some interim notes.
I suspected this may have relation to rebooting and restart of node, but that appeared wrong lead, as error occurred multiple minutes into running state.
Next, let's look into relevant code in libndctl which is `ndctl_region_get_pfn_seed`:
 "pfn: no seed" is reported in 3 cases, we can leave out 1st as its very unlikely and should log a message.
Remaining cases are:
1. can't read /sys/class/nd/ndctl0/device/region0/pfn_seed
2. content of file read in 1. does not match value returned by `ndctl_pfn_get_devname`, which seems to return a string based on internal context state (did I get that right?)

Quicker approach is to examine pfn_seed contents, so I tried that by tracing the presence and contents of `/sys/class/nd/ndctl0/device/region0/pfn_seed` on a node (note that this is done in node context, not in container context where pmem-csi runs).
It appears the file is there and readable all the time, so we can rule out 1, leaving choice 2.
What is interesting to note that the contents of /sys/class/nd/ndctl0/device/region0/pfn_seed
fluctuate (slowly, few minutes interval) between `pfn0.0` and `pfn0.1` while the testing is running, so if the internal state
returned by `ndctl_pfn_get_devname` does not match, it could produce the error we see.

The next step needed is to instrument libndctl with some added tracing to see:
1. what are the pfn_seed values it compares
2. where is internal state of pfn_seed value created and when changed

Also, worth checking:
1. do we see exact same state in host context and container context?
2. why does pfn_seed value is /sys fluctuate, is this correct and expected, or is this sign of some problem?

**Comment by pohly** [comment]
Olev Kartau <notifications@github.com> writes:
> Also, worth checking:
> 1. do we see exact same state in host context and container context?
> 2. why does pfn_seed value is /sys fluctuate, is this correct and
> expected, or is this sign of some problem?

Can you ask the Intel kernel developers these questions and also explain
the problem to them?

**Comment by okartau** [comment]
Current theory is, when we run a new binary during testing in a pod, it creates another copy of libndctl with new context. Later the original pod libndctl context gets used again, and may conflict with made changes.
Pmem and /sys file system are host-global, means we should not have more than one libndctl context per host.

**Comment by okartau** [comment]
i.e, we can not run any device-modifying tests in direct mode that bypass the original pod device manager, right?

**Comment by pohly** [comment]
Olev Kartau <notifications@github.com> writes:
> Current theory is, when we run a new binary during testing in a pod,
> it creates another copy of libndctl with new context. Later the
> original pod libndctl context gets used again, and may conflict with
> made changes.  Pmem and /sys file system are host-global, means we
> should not have more than one libndctl context per host.

That's not something that we can enforce. It would also prevent an admin
from doing anything with PMEM while PMEM-CSI is running. That would
negate one of the advantages of "direct" mode, the one that we don't
reserve the entire space exclusively for PMEM-CSI.

Instead, each new operation must start with a fresh libndctl context (if
that is the problem).

**Comment by pohly** [comment]
Olev Kartau <notifications@github.com> writes:
> i.e, we can not run any device-modifying tests in direct mode that
> bypass the original pod device manager, right?

We should be ablt to run the tests in parallel. That's the same as an
admin logging into the node and making some changes.

**Comment by okartau** [comment]
> We should be able to run the tests in parallel.

That's good desired state, but reality now is that we can not keep two contexts open at same time.
This is acknowledged by our own comment in pmd-ndctl.go where we point to issue 96 of ndctl (still open).
In one context, we use mutex to keep things safe.
If we want to run in parallel, we either have to expand that mutex to cover those parallel executions, or to destroy and start context every time we use it, means we cant keep the original pod running and assuming it has valid context.
Admin cant log in and change the pmem state at same time when driver has context open, thats limitation documented by issue 96.

**Comment by pohly** [comment]
What is this "context"? Why can't we reinitialize libndctl each time we use it?

**Comment by pohly** [comment]
https://github.com/pmem/ndctl/issues/96 is about a Linux kernel panic. Therefore I am not sure how this relates to this issue here.

**Comment by pohly** [comment]
> What is this "context"?

Probably https://github.com/intel/pmem-csi/blob/1ead73ace917ee10059617781d7e0d958a665a3f/pkg/pmem-device-manager/pmd-ndctl.go#L20

It's not hard to make that a local variable that the different methods allocate anew when they need it.

If the root cause is that this context caches some information, then we reduce the conflict from "don't do anything while PMEM-CSI runs" to "don't do anything while PMEM-CSI is actively doing something" - not perfect, but a lot better.

**Comment by okartau** [comment]
> [pmem/ndctl#96](https://github.com/pmem/ndctl/issues/96) is about a Linux kernel panic. Therefore I am not sure how this relates to this issue here.

True, I think I was mixing those without proof they are related.
Our use case with multiple programs linked with libndctl is different from multiple invocations of ndctl, and also the trouble we hit is different.

**Comment by okartau** [comment]
With "context" here I refer to part that causes "no seed" failure we see, i.e. pfn->pfn_seed which is a string in libndctl internal memory structure, which seems to get allocated anew for a new instance, so another instance cant reuse it but gets own copy.

**Comment by okartau** [comment]
> reinitialize libndctl each time we use it?

Good idea! We should investigate how that works and what is the overhead. Separate libndctl sessions will make the critical regions much shorter.
However there is still no global guard (like host level mutex) that would guarantee that multiple operations do not try changes at same time. I think with that style we will get to the level where that issue 96 gets closer. However, the probability to hit trouble in practice is not high.

**Comment by pohly** [comment]
> It's not hard to make that a local variable that the different methods allocate anew when they need it.

See PR #488.

> what is the overhead

The only effect would be on "number of volume changes per second", which isn't a relevant criteria for PMEM-CSI.

> I think with that style we will get to the level where that issue 96 gets closer.

Why closer? PMEM-CSI still has the mutex lock for its own operations, so the risk is the same as before.

**Comment by okartau** [comment]
> Why closer?

What I mean is that by creating a context for every operations, our use of ndctl device will be close to pattern of calling ndctl CLI utility each time. And repeating parallel calling of ndctl can lead to kernel panic as in issue 96. But that is still practical improvement over current state where we keep context opened forever in one process.

**Comment by pohly** [comment]
> What I mean is that by creating a context for every operations, our use of ndctl device will be close to pattern of calling ndctl CLI utility each time.

I doubt that this issue 96 was caused by each ndctl instance creating a new context. It's much more likely that it was the number of actual changes requested from the kernel.

**Comment by okartau** [comment]
Closing, should be resolved via #488
